metric_learn/mmc.py

```diff
@@ -344,10 +340,11 @@ def _D_constraint(self, X, c, d, w):
     diff = X[c] - X[d]
     diff_sq = diff * diff
     dist = np.sqrt(diff_sq.dot(w))
-    sum_deri1 = np.sum(diff_sq / (2 * np.maximum(dist, 1e-6))[:,None], axis = 0)
-    sum_deri2 = np.sum(
-      np.einsum('ij,ik->ijk', diff_sq, diff_sq) / (-4 * np.maximum(1e-6, dist**3)[:,None,None]),
-      axis = 0
+    sum_deri1 = np.einsum('ij,i', diff_sq, 0.5 / np.maximum(dist, 1e-6))
+    sum_deri2 = np.einsum(
+      'ijk,i',
+      np.einsum('ij,ik->ijk', diff_sq, diff_sq),
+      -0.25 / np.maximum(1e-6, dist**3)
```

Comment:
Nested calls to `np.einsum` are pretty hard to read. I think the previous version was fine, but you could also pull out the inner `einsum` to a local variable if this version is significantly faster.

It's also possible to do `einsum` with >2 arrays, but that can have unintended consequences (see https://github.com/numpy/numpy/issues/5366). As of numpy 1.12 the situation is a bit better (see https://github.com/numpy/numpy/pull/5488), but I don't think we need to go down that route.